### PR TITLE
Add regression test for regexp_replace hanging with some inputs

### PR DESCRIPTION
**Pull request by andygrove**
Closes https://github.com/NVIDIA/spark-rapids/issues/8323 and https://github.com/NVIDIA/spark-rapids/issues/8448

This PR adds a regression test for an issue that was [recently fixed](https://github.com/rapidsai/cudf/issues/13404) in cuDF. This test uncovered another bug (https://github.com/NVIDIA/spark-rapids/issues/8448), which is now also fixed in this PR.

### CODE REVIEW

**Review by gerashegalov — DISMISSED** [review]

File: integration_tests/src/main/python/regexp_test.py [hunk 1]
```diff
@@ -956,3 +956,11 @@ def test_regexp_memory_ok():
             'spark.rapids.sql.batchSizeBytes': '20' # 1 row in the batch
         }
     )
+
+def test_re_replace_all():
+    # regression test for https://github.com/NVIDIA/spark-rapids/issues/8323
```
Comment (mythrocks) [inline integration_tests/src/main/python/regexp_test.py:961]: Nitpick: Do we have a prescribed style for function documentation? I've been using the following even for tests:
```suggestion
    """
    regression test for https://github.com/NVIDIA/spark-rapids/issues/8323
    """
```

**Comment by andygrove** [comment]
build

File: integration_tests/src/main/python/regexp_test.py (hunk 1 above)
Comment (andygrove) [reply integration_tests/src/main/python/regexp_test.py]: Good point. Probably best to stick with the Python documentation approach.

**Comment by andygrove** [comment]
build

**Comment by andygrove** [comment]
The new test fails against Spark 3.1.1 due to CPU and GPU producing different results. It looks like behavior is different between Spark 3.1.1 and 3.3.2 and the GPU matches the behavior in 3.3.2. I am investigating.

**Comment by andygrove** [comment]
> The new test fails against Spark 3.1.1 due to CPU and GPU producing different results. It looks like behavior is different between Spark 3.1.1 and 3.3.2 and the GPU matches the behavior in 3.3.2. I am investigating.

I filed https://github.com/NVIDIA/spark-rapids/issues/8448

**Comment by andygrove** [comment]
build

**Comment by NVnavkumar** [comment]
I think the documentation will need to be updated here: https://github.com/NVIDIA/spark-rapids/blob/1b93b62cdea6709e090a12e92ed6e6f0436f9dcb/docs/compatibility.md?plain=1#L450 since we actually do support replace with backrefs.

**Review by NVnavkumar — APPROVED** [review]

**Review by mythrocks — APPROVED** [review]
LGTM!

**Pull request merged**